{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Input data preparation\n",
    "This notebooks prepares the data files needed for the cell-type specific gene regulatory network (GRN) inference pipeline.\n",
    "## Preparation of individual input files\n",
    "This section separately prepares each input file/folder as subsections. In each subsection, we will describe the expected input file, demonstate the preparation script with usage displayed when available, and briefly illustrate the content and/or format of the prepared input file. All these input files are placed in the `data` folder of this inference pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "dictys_data_path = '/ocean/projects/cis240075p/asachan/datasets/B_Cell/multiome_1st_donor_UPMC_aggr/dictys_outs/data'\n",
    "multiome_data_path = '/ocean/projects/cis240075p/asachan/datasets/B_Cell/multiome_1st_donor_UPMC_aggr/outs/filtered_feature_bc_matrix/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The helper function expression_mtx.py can also take in multiomic cell-ranger arc produced features that have two categories 'Gene Expression' and 'Peaks' and filter out all peak names and gene names which have : and . in them {peak name is of the type chrN:start-end}\n",
    "#### I moved the expression.tsv file from htc_crc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "gzip: stdout: Broken pipe\n",
      "cat: write error: Broken pipe\n",
      "            AAACAGCCAAACCTTG-1  AAACAGCCAAAGCTAA-1  AAACAGCCAAGCCACT-1\n",
      "A1BG                         0                   0                   0\n",
      "A1BG-AS1                     0                   0                   0\n",
      "A1CF                         0                   0                   0\n",
      "A2M                          0                   0                   0\n"
     ]
    }
   ],
   "source": [
    "# read and print head of a .tsv.gz file from the dictys_data_path/expression.tsv.gz\n",
    "!printf '%-10s%20s%20s%20s\\n' '' $(cat /ocean/projects/cis240075p/asachan/datasets/B_Cell/multiome_1st_donor_UPMC_aggr/dictys_outs/data/expression.tsv.gz | gunzip | head -n 5 | awk -F \"\\t\" '{print $1\"\\t\"$2\"\\t\"$3\"\\t\"$4}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24037\n"
     ]
    }
   ],
   "source": [
    "#print the number of lines in the .tsv.gz file\n",
    "!zcat $dictys_data_path/expression.tsv.gz | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sort your bams to get aligned reads per cell (36k bam files)\n",
    "#### a. Submitted array jobs to sort time-point wise bams"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### b. Subset your cell barcodes in the clusters you want the GRNs to be calculated for (cell-types from ledien clustering in the aggregated anndata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "adata=sc.read_h5ad('/ocean/projects/cis240075p/asachan/datasets/B_Cell/multiome_1st_donor_UPMC_aggr/outs/adata_aggregated_gene.leiden.h5ad')\n",
    "adata.obs.leiden.to_csv(f'{dictys_data_path}/clusters.csv',header=['Cluster'],index=True,index_label='Barcode') #12 leiden cluster labels in the 33k cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(0, 117)\\t2.0468006\\n  (0, 207)\\t2.0468006\\n...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(0, 4)\\t0.7918985\\n  (0, 16)\\t0.7918985\\n  (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(0, 26)\\t1.1051886\\n  (0, 35)\\t1.1051886\\n  ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(0, 10)\\t1.6259605\\n  (0, 38)\\t1.6259605\\n  ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(0, 10)\\t1.6710577\\n  (0, 12)\\t1.6710577\\n  ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32413</th>\n",
       "      <td>(0, 38)\\t1.7591833\\n  (0, 100)\\t2.3623042\\n ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32414</th>\n",
       "      <td>(0, 60)\\t1.9571365\\n  (0, 100)\\t1.9571365\\n ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32415</th>\n",
       "      <td>(0, 3)\\t0.6992017\\n  (0, 12)\\t0.6992017\\n  (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32416</th>\n",
       "      <td>(0, 5)\\t0.8446401\\n  (0, 22)\\t0.8446401\\n  (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32417</th>\n",
       "      <td>(0, 26)\\t0.6173298\\n  (0, 29)\\t0.6173298\\n  ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32418 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       0\n",
       "0        (0, 117)\\t2.0468006\\n  (0, 207)\\t2.0468006\\n...\n",
       "1        (0, 4)\\t0.7918985\\n  (0, 16)\\t0.7918985\\n  (...\n",
       "2        (0, 26)\\t1.1051886\\n  (0, 35)\\t1.1051886\\n  ...\n",
       "3        (0, 10)\\t1.6259605\\n  (0, 38)\\t1.6259605\\n  ...\n",
       "4        (0, 10)\\t1.6710577\\n  (0, 12)\\t1.6710577\\n  ...\n",
       "...                                                  ...\n",
       "32413    (0, 38)\\t1.7591833\\n  (0, 100)\\t2.3623042\\n ...\n",
       "32414    (0, 60)\\t1.9571365\\n  (0, 100)\\t1.9571365\\n ...\n",
       "32415    (0, 3)\\t0.6992017\\n  (0, 12)\\t0.6992017\\n  (...\n",
       "32416    (0, 5)\\t0.8446401\\n  (0, 22)\\t0.8446401\\n  (...\n",
       "32417    (0, 26)\\t0.6173298\\n  (0, 29)\\t0.6173298\\n  ...\n",
       "\n",
       "[32418 rows x 1 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.DataFrame(adata.X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "leiden\n",
       "0     5283\n",
       "1     4647\n",
       "2     4440\n",
       "3     4365\n",
       "4     3984\n",
       "5     2820\n",
       "6     2515\n",
       "7     2345\n",
       "8      999\n",
       "9      657\n",
       "10     258\n",
       "11     105\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.leiden.value_counts() #number of cells in each cluster"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Submit bash script to get subsets folders with the barcode names per cell-type for rna and atac data. input is clusters.csv (coming from the aggr_anndata's leiden clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subset0\n",
      "Subset1\n",
      "Subset2\n",
      "Subset3\n",
      "Subset4\n",
      "Subset5\n",
      "Subset6\n",
      "Subset7\n",
      "Subset8\n",
      "Subset9\n",
      "AAACAGCCAAGGTGCA-1\n",
      "AAACATGCAAAGCTCC-1\n",
      "AAACATGCACAGAAAC-1\n",
      "AAACATGCAGAGAGCC-1\n",
      "AAACAGCCAAGGTGCA-1\n",
      "AAACATGCAAAGCTCC-1\n",
      "AAACATGCACAGAAAC-1\n",
      "AAACATGCAGAGAGCC-1\n"
     ]
    }
   ],
   "source": [
    "################# Check the subsets output #################\n",
    "#Cell subset list\n",
    "!head $dictys_data_path/subsets.txt\n",
    "# #RNA cell barcodes for Subset 1\n",
    "!head -n 4 $dictys_data_path/subsets/Subset1/names_rna.txt\n",
    "#ATAC cell barcodes for Subset 1. They are identical because it's a joint profiling dataset.\n",
    "!head -n 4 $dictys_data_path/subsets/Subset1/names_atac.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Use motifs from HOCOMOCO (wget-sbatch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">dKhGCGTGh\tAHR_HUMAN.H11MO.0.B\t3.3775000000000004\n",
      "0.262728374765856\t0.1227600511842322\t0.362725638699551\t0.25178593535036087\n",
      "0.07633328991810645\t0.08258130543118362\t0.22593295481662123\t0.6151524498340887\n",
      "0.14450570038747923\t0.28392173880411337\t0.13815442099009081\t0.4334181398183167\n",
      "0.023935814057894068\t0.016203821748029118\t0.9253278681170539\t0.03453249607702277\n",
      "0.007919544273173793\t0.953597675415874\t0.017308392078009837\t0.021174388232942286\n",
      "0.02956192959210962\t0.012890110758086997\t0.9474192747166682\t0.010128684933135217\n",
      "0.007919544273173797\t0.029561929592109615\t0.012337825593096645\t0.9501807005416201\n",
      "0.007919544273173793\t0.007919544273173793\t0.9762413671804787\t0.007919544273173793\n",
      "0.27886589130660366\t0.4285328543459993\t0.10955683916661985\t0.18304441518077724\n",
      ">hnnGGWWnddWWGGdbWh\tAIRE_HUMAN.H11MO.0.C\t5.64711\n",
      "0.38551919443239085\t0.2604245534178759\t0.1353299124033618\t0.21872633974637148\n",
      "0.18745267949274294\t0.18745267949274294\t0.14575446582123766\t0.4793401751932764\n",
      "0.14575446582123777\t0.14575446582123777\t0.2083017863284958\t0.5001892820290286\n",
      "0.0623580384782284\t0.020659824806723626\t0.8337749914010671\t0.08320714531398087\n",
      "0.020659824806723633\t0.020659824806723633\t0.7712276708938095\t0.18745267949274314\n",
      "0.3125473205072573\t0.062358038478228406\t0.04150893164247602\t0.5835857093720382\n",
      "0.37509464101451423\t0.10405625214973326\t0.04150893164247601\t0.4793401751932766\n"
     ]
    }
   ],
   "source": [
    "# see the output to check if gene names match TF names in anndata\n",
    "!head -n 18 $dictys_data_path/motifs.motif"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get the reference genome from homer directory - sbatch "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 4.4G\n",
      "4.0K annotations\n",
      " 12K chrom.sizes\n",
      "3.1G genome.fa\n",
      "3.2M hg38.aug\n",
      " 42M hg38.basic.annotation\n",
      "673M hg38.full.annotation\n",
      "164K hg38.miRNA\n",
      "505M hg38.repeats\n",
      " 24M hg38.rna\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#check the reference genome\n",
    "ls -h1s /ocean/projects/cis240075p/asachan/datasets/B_Cell/multiome_1st_donor_UPMC_aggr/dictys_outs/data/genome | head"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get gene gtf from ensembl - inline > extract genes in bed format - sbatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t11869\t14409\tDDX11L1\t.\t+\n",
      "chr1\t14404\t29570\tWASH7P\t.\t-\n",
      "chr1\t17369\t17436\tMIR6859-1\t.\t-\n",
      "chr1\t29554\t31109\tMIR1302-2HG\t.\t+\n",
      "chr1\t30366\t30503\tMIR1302-2\t.\t+\n",
      "chr1\t34554\t36081\tFAM138A\t.\t-\n",
      "chr1\t52473\t53312\tOR4G4P\t.\t+\n",
      "chr1\t57598\t64116\tOR4G11P\t.\t+\n",
      "chr1\t65419\t71585\tOR4F5\t.\t+\n",
      "chr1\t131025\t134836\tCICP27\t.\t+\n"
     ]
    }
   ],
   "source": [
    "!head $dictys_data_path/gene.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get blacklist regions of the genome from encode, to exclude them from the atac analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-08-16 17:57:21--  https://www.encodeproject.org/files/ENCFF356LFX/@@download/ENCFF356LFX.bed.gz\n",
      "Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144\n",
      "Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.\n",
      "HTTP request sent, awaiting response... 307 Temporary Redirect\n",
      "Location: https://encode-public.s3.amazonaws.com/2020/05/05/bc5dcc02-eafb-4471-aba0-4ebc7ee8c3e6/ENCFF356LFX.bed.gz?response-content-disposition=attachment%3B%20filename%3DENCFF356LFX.bed.gz&AWSAccessKeyId=ASIATGZNGCNXZD6QL5EC&Signature=%2BC%2B%2Bdz1d%2FFMSras1k46ru33iE9U%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEP7%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJGMEQCIBfRqv1kvRIi3OyVke5xBShuVEkO6EQDVQmyzUjBW2vQAiAi7Ra166Ijsiij%2BajjvzBxxXTwQAVgu86d%2ByENEMZUIiq8BQj3%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F8BEAAaDDIyMDc0ODcxNDg2MyIM22gFho2NRpSAVCAkKpAFZYQ2sb%2BUm%2BXqMdD6Q6PrIRUkJe6%2F15B%2FlWy%2F4GEOo0XMaXj8xK%2B5fIZMQS%2BaAjqchbByIJn9UBE6r%2FpO3s8vulrXIkLiwuoORYPRAeD0aKcftGhOhMor7vLEsli%2FTHf%2B3k0Gd9vWVZhDmGmxNPsIELBzVBOByCmOu%2F6vr0PhwPL64zUByVxjj2oVqRaMoS%2FoKlRO6%2Bw%2BGtN63srhZQtS1u6aClSeHUzFJgH9qshJRJWQkR49xrtTIt5GC5JmS5NtI4yBAxCobGzV0AzT0FHVW7ZbHJ1kMSjf40kWh283sVfeHrRdhZO2nPt13c9Iq3D%2FA4k2sT3GUROr8CQpZbcT7Bgi%2Bj5qPrVyU%2B2IRlVYAYqL9BbW1IgQANFGbiFfzxDOt7NyVKseR1YgLfz18pUCv4I9lBv3Ag%2B8ENHyn4DslrN4ZX7CdMFfgpC72ZS30WxSAaE2FmgIorOnBS2Sq3S6gYOvxnVps8Xxut0MlxFiRMi%2FbbsHmH5wE2uWH72NA%2FHy0UmClKBezeETi4BIllte0wieKN10tD%2FS2oMmux3bxxnI8T7Wh6e7P0%2Fir15duocnuNA%2Fmn0c%2BbqGWy3UBNUQ%2FPhmKkpuP96XaDIJD8SDW%2FY%2FtzAp0FJwNwN6ceNvPR%2BxOGrkPwBQ8kpk4u75iiUynNCKg3R0I6Fd0%2Fzt39tCZs3W4Rk2Coj1LuRpTS9ZpzpoNQxIfNiLTmu%2BPPTmSCOc5aBzWwAdJ91GPCPpMV%2F4tP5jaNVs%2BtpR%2BFvhVJJm8N9e1t7gCXWjk3WfX1sIcRQFfb7dQjnn%2FnvP8yLaNNtnTein97L27HxChcRr4jTeP8%2Bo9z88q%2Beom8%2FPRR%2B%2Bdom4fY9zAGgdAz%2FqKQZJssgkFhYw2JT%2FtQY6sgFLavOpEuwINrXUDUlTRG%2FIyg6rXZIH5mNGthz2JJliMCN90t1RYJvyfd1TX4Ni%2Fg3C4c0rBFChoeFFvOKoiGctj642cnBQQ41Gq1bmNCiYBcg96L%2FZ90E8oQNr%2B6X1knKCjrk%2FEIWKxJcKVcQ2%2F2wdazn4B1dnL7GAy6Vv4eIA%2FrrSUTiJPEkZL4jVMgdJOOc%2F34O1yBzlJcJhwexw4AyuveqMEAC%2BmK%2BJ0W7rsOESvf1H&Expires=1723975042 [following]\n",
      "--2024-08-16 17:57:22--  https://encode-public.s3.amazonaws.com/2020/05/05/bc5dcc02-eafb-4471-aba0-4ebc7ee8c3e6/ENCFF356LFX.bed.gz?response-content-disposition=attachment%3B%20filename%3DENCFF356LFX.bed.gz&AWSAccessKeyId=ASIATGZNGCNXZD6QL5EC&Signature=%2BC%2B%2Bdz1d%2FFMSras1k46ru33iE9U%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEP7%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJGMEQCIBfRqv1kvRIi3OyVke5xBShuVEkO6EQDVQmyzUjBW2vQAiAi7Ra166Ijsiij%2BajjvzBxxXTwQAVgu86d%2ByENEMZUIiq8BQj3%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F8BEAAaDDIyMDc0ODcxNDg2MyIM22gFho2NRpSAVCAkKpAFZYQ2sb%2BUm%2BXqMdD6Q6PrIRUkJe6%2F15B%2FlWy%2F4GEOo0XMaXj8xK%2B5fIZMQS%2BaAjqchbByIJn9UBE6r%2FpO3s8vulrXIkLiwuoORYPRAeD0aKcftGhOhMor7vLEsli%2FTHf%2B3k0Gd9vWVZhDmGmxNPsIELBzVBOByCmOu%2F6vr0PhwPL64zUByVxjj2oVqRaMoS%2FoKlRO6%2Bw%2BGtN63srhZQtS1u6aClSeHUzFJgH9qshJRJWQkR49xrtTIt5GC5JmS5NtI4yBAxCobGzV0AzT0FHVW7ZbHJ1kMSjf40kWh283sVfeHrRdhZO2nPt13c9Iq3D%2FA4k2sT3GUROr8CQpZbcT7Bgi%2Bj5qPrVyU%2B2IRlVYAYqL9BbW1IgQANFGbiFfzxDOt7NyVKseR1YgLfz18pUCv4I9lBv3Ag%2B8ENHyn4DslrN4ZX7CdMFfgpC72ZS30WxSAaE2FmgIorOnBS2Sq3S6gYOvxnVps8Xxut0MlxFiRMi%2FbbsHmH5wE2uWH72NA%2FHy0UmClKBezeETi4BIllte0wieKN10tD%2FS2oMmux3bxxnI8T7Wh6e7P0%2Fir15duocnuNA%2Fmn0c%2BbqGWy3UBNUQ%2FPhmKkpuP96XaDIJD8SDW%2FY%2FtzAp0FJwNwN6ceNvPR%2BxOGrkPwBQ8kpk4u75iiUynNCKg3R0I6Fd0%2Fzt39tCZs3W4Rk2Coj1LuRpTS9ZpzpoNQxIfNiLTmu%2BPPTmSCOc5aBzWwAdJ91GPCPpMV%2F4tP5jaNVs%2BtpR%2BFvhVJJm8N9e1t7gCXWjk3WfX1sIcRQFfb7dQjnn%2FnvP8yLaNNtnTein97L27HxChcRr4jTeP8%2Bo9z88q%2Beom8%2FPRR%2B%2Bdom4fY9zAGgdAz%2FqKQZJssgkFhYw2JT%2FtQY6sgFLavOpEuwINrXUDUlTRG%2FIyg6rXZIH5mNGthz2JJliMCN90t1RYJvyfd1TX4Ni%2Fg3C4c0rBFChoeFFvOKoiGctj642cnBQQ41Gq1bmNCiYBcg96L%2FZ90E8oQNr%2B6X1knKCjrk%2FEIWKxJcKVcQ2%2F2wdazn4B1dnL7GAy6Vv4eIA%2FrrSUTiJPEkZL4jVMgdJOOc%2F34O1yBzlJcJhwexw4AyuveqMEAC%2BmK%2BJ0W7rsOESvf1H&Expires=1723975042\n",
      "Resolving encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)... 52.218.250.99, 52.218.229.187, 52.92.233.49, ...\n",
      "Connecting to encode-public.s3.amazonaws.com (encode-public.s3.amazonaws.com)|52.218.250.99|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 8211 (8.0K) [application/vnd.realvnc.bed]\n",
      "Saving to: ‘/ocean/projects/cis240075p/asachan/datasets/B_Cell/multiome_1st_donor_UPMC_aggr/dictys_data/blacklist.bed.gz’\n",
      "\n",
      "/ocean/projects/cis 100%[===================>]   8.02K  --.-KB/s    in 0s      \n",
      "\n",
      "2024-08-16 17:57:23 (333 MB/s) - ‘/ocean/projects/cis240075p/asachan/datasets/B_Cell/multiome_1st_donor_UPMC_aggr/dictys_data/blacklist.bed.gz’ saved [8211/8211]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://www.encodeproject.org/files/ENCFF356LFX/@@download/ENCFF356LFX.bed.gz -O $dictys_data_path/blacklist.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!gunzip $dictys_data_path/blacklist.bed.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t628903\t635104\n",
      "chr1\t5850087\t5850571\n",
      "chr1\t8909610\t8910014\n",
      "chr1\t9574580\t9574997\n",
      "chr1\t32043823\t32044203\n",
      "chr1\t33818964\t33819344\n",
      "chr1\t38674335\t38674715\n",
      "chr1\t50017081\t50017546\n",
      "chr1\t52996949\t52997329\n",
      "chr1\t55372488\t55372869\n"
     ]
    }
   ],
   "source": [
    "!head $dictys_data_path/blacklist.bed"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dictys",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
